Jetpack Section: Restore - Replace Rewind Status Service with Restore Use Cases

ParaskP7 opened the pull request:
Parent #13328

This PR replaces the `RewindStatusService` with the `PostRestoreUseCase` and `GetRestoreStatusUseCase`. This step is necessary in order to reuse those use cases, remove the complicated status services and in general reduce the complexity within the `ActivityLogViewModel`. Also, this step is a prerequisite and as much necessary for another feature that is about to be added to this screen, which is showing a progress row when a `Backup Download` has been triggered.

⚠️ When this PR is merged into `develop` is will affect the next release as this refactor is not guarded on a feature flag. Make sure to do a thorough code review. @zwarm @malinajirka and @ashiagr I have added all of you as reviewers to this PR as this is a complicated one and I would love to get as much comments as possible. At least, testing the PR and finding out bugs would be a major help.

ℹ️ Don't be scared by the amount of commits (77), the file changes are not that many (9). The commits are that many because of the many refactors I had to do in order to improve the state of the `ActivityLogViewModel` and its corresponding tests to be able to remove the status service and replace that with the use cases. If you do this code review commit-by-commit it will help you enormously.

Future TODOs:
* Timestamp is not being shown on restore progress row and its corresponding snackbars (started and completed) when triggered from the web. An item for that will be added on the parent ticket.
* `RewindStatusService` has not been completely removed due to the fact that the `ActivityLogDetailViewModel` also uses it just to show/hide the `RESTORE` button. An item for that will be added on the parent ticket.

To test:

#### Scenario.1 (App flow)
* Open app and go to `My Site` tab,
* Launch `Activity Log` screen,
* Find a rewindable item from the list and: 
  * Either, click on the `RESTORE` icon and then on the dialog that appears, click on the `RESTORE SITE` button,
  * Or, click on the item itself to launch the `Activity Log Details` screen, click the `RESTORE` button and then on the dialog that appears, click on the `RESTORE SITE` button,
* You should automatically be navigated back to the `Activity Log` screen, on top of the list, and a restore will be triggered,
* The restore progress item should immediately appears. Make sure the restore timestamp is available on this progress item.
* A snackbar should immediately appear informing that the restore just started. Make sure the restore timestamp is available on the snackbar.
* Make sure the `RESTORE` icon, for every rewindable item, is not shown while restore progress is ongoing.
* Wait for restore to finish while having the `Activity Log` screen open both, within the app and the web (so you can see the actual progress),
* When restore finishes (looking on the web), a the restore (push) notification should appear.
* Then, the restore progress item should disappear (after at least 5'' seconds, due to the while loop check status delay).
* A pull-to-refresh should be automatically triggered and a new restore item might (or might not) appear (depending on how fast the restore progress item disappears and how fast this new item becomes available on the backend, via the `activityLogStore.fetchActivities(...)` API call). In case the new restore item does not appear, wait a bit (maybe 10'' seconds) and trigger a manual pull-to-refresh. Verify that the new restore item appears after some time.
* A snackbar should immediately appear informing that the restore just completed.  Make sure the restore timestamp is available on the snackbar.
* Make sure the `RESTORE` icon, for every rewindable item, is now shown again.

#### Scenario.2 (Web flow)
* Open app and go to `My Site` tab,
* Trigger restore from web,
* Wait for it to start,
* Either, launch `Activity Log` screen, or trigger a pull-to-refresh if you are already within that screen.
* The restore progress item should immediately appears. **CHANGE** (⚠️): The timestamp will not be available when triggered from the web, or it might become available based on some constraints (to be improved).
* Make sure the `RESTORE` icon, for every rewindable item, is not shown while restore progress is ongoing.
* Wait for restore to finish while having the `Activity Log` screen open both, within the app and the web (so you can see the actual progress),
* When restore finished (looking on the web), a the restore (push) notification should appear.
* Then, the restore progress item should disappear (after at least 5'' seconds, due to the while loop check status delay).
* A pull-to-refresh should be automatically triggered and a new restore item might (or might not) appear (depending on how fast the restore progress item disappears and how fast this new item becomes available on the backend, via the `activityLogStore.fetchActivities(...)` API call). In case the new restore item does not appear, wait a bit (maybe 10'' seconds) and trigger a manual pull-to-refresh. Verify that the new restore item appears after some time.
* A snackbar should immediately appear informing that the restore just completed. **CHANGE** (⚠️): The timestamp will not be available when triggered from the web, or it might become available based on some constraints (to be improved).
* Make sure the `RESTORE` icon, for every rewindable item, is now shown again.

#### Scenario.3 (Go wild flow)
* Trigger any of the above 2 scenarios until you reach the point where the restore progress item is shown.
* Trigger a pull-to-refresh manually and make sure the progress item is still shown.
* Go to the bottom of the list and trigger another page load, go to the top of the list and make sure the progress item is still shown.
* Go to the bottom of the list and trigger another page load, click on any item to navigate to the `Activity Log Detail` page, and then go back. Go to the top of the list and make sure the progress item is still shown.
* Go to the bottom of the list and trigger another page load, wait there until the restore finishes.
* Then, make sure you are being navigated automatically to the top of the list and the restore progress item have disappeared.

#### Scenario.4 (Restore flow)
* Since this PR updates the `GetRestoreStatusUseCase` it is recommended that the `Restore Flow` is tested as well.
* To do so, open app and go to `My Site` tab,
* Go to `My Site` -> `Toolbar Avatar` -> `App Settings` -> `Test feature configuration`.
* Enable `RestoreFeatureConfig`, scroll down and click the `RESTART THE APP` button. 
* Launch `Activity Log` screen,
* Find a rewindable item from the list, click on the ellipsis menu item and select `Restore to this point` to launch the `Restore screen,
* Click the `Restore site` button. On the confirmation screen, click on the `Confirm restore site` button.
* Wait for the restore to finish, click `Done` and in general verify everything works as expected.

PS: I **REALLY** suggest doing a commit-by-commit review as this will make it much easier for you to review the overall solution.

PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ParaskP7 committed:
Feature: Make post restore use case types parameter nullable

This is required in order to allow running this use case from within the
activity log screen when that screen is launched with a rewind request
from the activity log detail screen, which doesn't have any 'type'
selection functionality.

See 'ActivityLogViewModel.onRewindConfirmed(...)' which only holds a
reference to the 'rewindId' and doesn't know anything about the 'types'.

ParaskP7 committed:
Refactor: Reformat get restore status function parameters

ParaskP7 committed:
Feature: Make get restore status use case restore id parameter nullable

This is required in order to allow running this use case from within the
activity log screen when a web restore launch is triggered and the user
(re)launching the screen or triggered a pull-to-refresh. When that
happens the activity log screen will query for any restore that might be
happening in order to show the progress row.

Thus, this 'restoreId' needs to become optional for when that id is not
known in advance. And based on that change the if logic to allow
querying for status needs to be updated to allow this scenario.

ParaskP7 committed:
Feature: Emit progress when the rewind status is queued

This is necessary in order to show progress before the actual rewind
running happens. Without that in place, if the user would click the
'Ok, notify me!' button too quick from within the restore screen, then
the progress row will not appear immediately on the activity log screen.

ParaskP7 committed:
Feature: Launch fetch activities rewind when restore id is null

Again, this is required in order to allow running this use case from
within the activity log screen when a web restore launch is triggered
and the user (re)launching the screen or triggered a pull-to-refresh.
When that happens the activity log screen will query for any restore
that might be happening in order to show the progress row.

However, unless the 'ActivityLogStore.fetchActivitiesRewind(...)' is
launched at least once 'ActivityLogStore.getRewindStatusForSite(...)'
will not get a valid 'rewind' object back. Thus, this fetching of the
activities rewind is necessary for that scenario to work properly.

ParaskP7 committed:
Refactor: Extract fetch activities rewind extension function

This is done in order to remove the corresponding duplication with the
fetch activities rewind triggering and error handling logic.

ParaskP7 committed:
Refactor: Change map not null to map for current activity type filter

This filter doesn't allow any nullable 'ActivityTypeModel' object and as
such the 'mapNotNull { ... }' function is not necessary since it can be
replaced with plain 'map { ... }'.

ParaskP7 committed:
Refactor: Optimise activity log list status imports

ParaskP7 committed:
Refactor: Optimise rewind status imports

ParaskP7 committed:
Refactor: Optimise activity log list item imports

ParaskP7 committed:
Refactor: Optimise activity log navigation events imports

ParaskP7 committed:
Test: Optimise rewind state test imports

ParaskP7 committed:
Test: Optimise rewind status service test imports

ParaskP7 committed:
Test: Optimise activity log action test imports

ParaskP7 committed:
Test: Add named parameters for activity log list item event on tests

ParaskP7 committed:
Test: Refactor activity log view model test setup function

ParaskP7 committed:
Test: Fix activity log view model test assert fetch events assertions

ParaskP7 committed:
Refactor: Remove unnecessary get rewind status on start store call

ParaskP7 committed:
Test: Refactor all tests to a single line arrange, act, assert structure

This is done in order for the test to be read with the minimum of effort
in terms of understanding where the arrange, act and assert related code
is.

ParaskP7 committed:
Test: Move private test function to the end and create private group

ParaskP7 committed:
Test: Extract first, second and third activity log list item functions

ParaskP7 committed:
Test: Remove unused rewind status model field

ParaskP7 committed:
Test: Optimise org wordpress android view model event import

ParaskP7 committed:
Test: Restructure and group activity log view model test fields

ParaskP7 committed:
Test: Add named parameters to initialize activity list activity variable

ParaskP7 committed:
Test: Extract helper activity published time function

ParaskP7 committed:
Test: Reformat expected activity list helper function

ParaskP7 committed:
Test: Rename activity log list items variable to plain list

ParaskP7 committed:
Test: Reorder helper functions based on usage

ParaskP7 committed:
Test: Improve loads next page on scroll to bottom test

ParaskP7 committed:
Test: Extract helper activity function and replace all activity usages

ParaskP7 committed:
Test: Replace run blocking with internal helper test coroutine function

ParaskP7 committed:
Test: Extract helper event function and replace all activity usages

ParaskP7 committed:
Refactor: Use mutable live data only where required

ParaskP7 committed:
Test: Add reload events tests for activity log view model

In order to be able to add 'reloadEvents(...)' specific tests to the
'ActivityLogViewModelTest' suite that functioned had to become
'@VisibleForTesting'. Otherwise it would have been to complicate to
create those tests.

Note that this exact need indicates that the 'reloadEvents(...)' could
be extracted to its own class in the future.

ParaskP7 committed:
Test: Add named parameters to activity log list item event function

ParaskP7 committed:
Test: Improve activity list init on activity log view model test

ParaskP7 committed:
Refactor: Replace disable actions display progress with restore event

This new 'RestoreEvent' data class will hold all the necessary
information that 'reloadEvents(...)' will need to function.

This refactor actually break the functionality since 'isCompleted' is
never set to true, but tests work as expected. This is a necessary step
for further refactorings, which would ultimately help replace the
'RewindStatusService' with the 'GetRestoreStatusUseCase'.

ParaskP7 committed:
Refactor: Reformat long function parameter to separate lines

ParaskP7 committed:
Refactor: Replace rewind progress with restore event rewind id

This refactor actually break the functionality since 'rewindId' is
never set to a value, but tests work as expected. This is a necessary
step for further refactorings, which would ultimately help replace the
'RewindStatusService' with the 'GetRestoreStatusUseCase'.

ParaskP7 committed:
Refactor: Replace rewinding activity with activity log item by rewind id

Instead of depending on the 'RewindStatusService' to get the
'rewindingActivity', this change makes it so that a store call
to 'getActivityLogItemByRewindId(...)' retrieves the rewinding activity.
Thus, this makes the whole view model less stateless and dependant on
the 'RewindStatusService'.

ParaskP7 committed:
Refactor: Replace rewinding activity with restore event date

Instead of depending on the 'RewindStatusService' to get the
'rewindingActivity', this change makes it so that the restore event will
hold the date info. Thus, this makes the whole view model less stateless
and dependant on the 'RewindStatusService'.

This refactor actually break the functionality since 'date' is
never set, but tests work as expected. This is a necessary step
for further refactorings, which would ultimately help replace the
'RewindStatusService' with the 'GetRestoreStatusUseCase'.

ParaskP7 committed:
Feature: Remove rewind status service from activity log view model

ParaskP7 committed:
Merge: Fix rebase conflict related test imports

ParaskP7 committed:
Refactor: Replace scoped view model with view model scope extension func

This is done for two main reasons:
1) Using Google's pure 'ViewModel' with its accompanying
'viewModelScope' KTX extension function is the recommended way, and
2) The custom 'ScopedViewModel' is using a 'Job', while 'viewModelScope'
is using a 'SupervisorJob'. The main difference between the 'Job' and
the 'SupervisorJob' is that children of a supervisor job can fail
independently of each other.

The 2nd reason above is really important for the next commit, which will
introduce another 'restore status job', in addition to the already
existing 'fetch activities job'. This new job will also be cancelled
every time before it being launched. Thus, using the current
'ScopedViewModel + Job' combination will cause the 'fetch activities
job' to cancel as well. By replacing that with 'viewModelScope +
SupervisorJob' this limitation is lifted and everything will work as
expected.

ParaskP7 committed:
Feature: Add restore use cases to activity log view model

These use cases are replacing the previously removed
'RewindStatusService'.

ParaskP7 committed:
Test: Move rewind confirmed test to its own dedicated section

ParaskP7 committed:
Test: Rename rewind confirmed related test to given, when, then pattern

ParaskP7 committed:
Test: Optimise rewind confirmed related test

ParaskP7 committed:
Test: Rename currently restoring activity log view mode test constant

ParaskP7 committed:
Feature: Add current restore event state on view model

Without this state when pull-to-refresh is triggered or any other such
trigger, which is based on 'requestEventsUpdate(...)' the restore event
would default to the wrong state.

ParaskP7 committed:
Feature: Enhance get restore status use case with date retrieval

This date will then be used as an addition to the 'rewindId' to for the
restore progress item and the restore finished message. That is in case
retrieving the 'rewindId' is not possible at that point through
'activityLogStore.getActivityLogItemByRewindId(...)'.

ParaskP7 committed:
Feature: Enhance handle restore status with date info

This date is used as an addition to the 'rewindId' to for the restore
progress item and the restore finished message. That is in case
retrieving the 'rewindId' is not possible at that point through
'activityLogStore.getActivityLogItemByRewindId(...)'.

ParaskP7 committed:
Feature: Reduce restore status use case delay from ten to five seconds

ParaskP7 committed:
Refactor: Extract emit progress function for restore status use case

ParaskP7 committed:
Refactor: Extract emit complete function for restore status use case

ParaskP7 committed:
Refactor: Extract emit failure function for restore status use case

ParaskP7 committed:
Refactor: Extract emit finished function for restore status use case

ParaskP7 committed:
Refactor: Remove braces from when entry for restore status use case

ParaskP7 committed:
Refactor: Extract is network unavailable function for restore use case

ParaskP7 committed:
Feature: Add initial is network unavailable check when restore is null

ParaskP7 committed:
Test: Add get restore status use case test suite

ParaskP7 committed:
Refactor: Rename is rewind to restore progressI item shown

ParaskP7 committed:
Refactor: Move on activity fetched close to usage

ParaskP7 committed:
Refactor: Move reload events close to first usage

ParaskP7 committed:
Refactor: Move request events update close to first usage

ParaskP7 committed:
Refactor: Move get rewind progress item close to usage

ParaskP7 committed:
Refactor: Rename get rewind to restore progress item

ParaskP7 committed:
Refactor: Move show rewind finished message close to usage

ParaskP7 committed:
Refactor: Simplify is restore progress item shown

ParaskP7 committed:
Refactor: Move on scrolled to bottom close to similar function

ParaskP7 committed:
Refactor: Rename on rewind to on restore confirmed

ParaskP7 committed:
Refactor: Rename show rewind to show restore started message

ParaskP7 committed:
Analysis: Fix ktlint issue with kotlin 1.4 trailing commas

ParaskP7 committed:
Analysis: Fix ktlint issue with test line length

ParaskP7 committed:
Analysis: Fix ktlint issue with kotlin 1.4 trailing commas

My previous commit missed those two cases in this commit because I
didn't understand that KtLint will actually fail when something like
that happens and not report the next issues on the same file. Thus, I
had to run it every time, fix the next issue, rerun, until KtLint
succeeds.

PS: Detekt will fix all these unnecessary problems with KtLint.

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/jetpack/restore/RestoreStates.kt:
```diff
@@ -103,10 +104,15 @@ sealed class RestoreRequestState {
         val message: String? = null,
-        val currentEntry: String? = null
+        val currentEntry: String? = null,
+        val date: Date? = null
```
🔍 Nitpick: We might want to consider renaming this value. The name doesn't tell us anything about the field except of its type. Is it current date or date of the last update of the progress or date when the action started?

ParaskP7 replied:
Your are totally right @malinajirka ! 🌟 

I am going to rename that to `published` and match that with the `ActivityLogModel`, since this is where that date comes from in the first place. Wdyt?

ParaskP7 replied:
This is done here: 4b5a9fda72cf636abcd3864be970fb1f02de7624

PS: Let me know if you want it renamed to something else. 🙏

malinajirka replied:
Awesome, thanks!

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/jetpack/restore/RestoreStates.kt:
```diff
@@ -103,10 +104,15 @@ sealed class RestoreRequestState {
+    data class Complete(
+        val rewindId: String,
+        val restoreId: Long,
+        val date: Date? = null
```
Same as above

ParaskP7 replied:
Same as above. 😄 

PS: Usually I try to avoid duplicate comment by adding a `PS: Xyz` on my first comment letting the author know that this comment applies elsewhere.

ParaskP7 replied:
This is done here: 4b5a9fd

malinajirka commented on WordPress/src/main/java/org/wordpress/android/viewmodel/activitylog/ActivityLogViewModel.kt:
```diff
@@ -177,24 +153,157 @@ class ActivityLogViewModel @Inject constructor(
         showFiltersIfSupported()
     }
 
+    @VisibleForTesting
```
🔍 Nitpick: In general we try to avoid using @VisibleForTesting as it's not necessary in most cases. The test would ideally test the state the class is providing through public methods/fields. Usually the need for using this annotation indicates that the code could/should be extracted into a separate class. Having said all that, feel free to keep it.

ParaskP7 replied:
@malinajirka 🙏 

I am so glad that you mentioned that and actually for saying that this is an indication that the code should be extracted into a separate class. This is exactly why I did that!

But, actually in this case I kind of had little luck in avoiding exposing the `reloadEvents(...)` since unfortunately I couldn't easily create all the combinations otherwise, in order to do the necessary refactoring, especially before adding the new solution.

Do you think now that a more simpler solution is in place I could possible make that back into a `private` function and simulate the exact some setup through the `public` entry points? I could probably achieve that but I might also miss some of it cases, not sure.

PS: I would personally prefer to leave it as such, hoping that later on it will be extracted out to a separate class.

malinajirka replied:
👍 Thanks for the clarification! I understand now, why/when it was added. I agree with you that keeping it as is, is good enough for now ;)!

ParaskP7 committed:
Refactor: Move flow on operator and dispatcher to restore use case

This is the most robust way of using this use case since clients do not
need to manage the dispatcher themselves. This way the use case will
always run in the background no matter which client will be using them.

Also, this change reduces some of the duplication where every client
needs to explicitly call the 'flowOn(...)' operator.

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/jetpack/restore/usecases/GetRestoreStatusUseCase.kt:
```diff
@@ -1,65 +1,100 @@
         }
+    }.flowOn(bgDispatcher)
+
+    private suspend fun FlowCollector<RestoreRequestState>.isNetworkUnavailable(): Boolean {
```
⚠️ We might want to consider renaming this method to "isNetworkAvailable". 
1. We use isNetworkAvailable throughout the codebase so the current name might lead to mistakes.
2. Both names isNetworkAvailable/isNetworkUnavailable are extremely similar so it's easy to miss the difference when scanning the code.

ParaskP7 replied:
I agree.

ParaskP7 replied:
This is done here: 830cc94e407f6313edd1bd97ac86fff9034db6f2

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/jetpack/restore/usecases/GetRestoreStatusUseCase.kt:
```diff
@@ -1,65 +1,100 @@
+    }.flowOn(bgDispatcher)
+
+    private suspend fun FlowCollector<RestoreRequestState>.isNetworkUnavailable(): Boolean {
+        if (!networkUtilsWrapper.isNetworkAvailable()) {
```
🔍 Nitpick: I like to have ideally a single return statement within a method when possible. We might want consider putting the return in front of the `if`. Wdyt?

ParaskP7 replied:
I kind of agree @malinajirka , for sure I understand why!

Would you prefer this version:

```
        return if (!networkUtilsWrapper.isNetworkAvailable()) {
            emit(NetworkUnavailable)
            true
        } else {
            false
        }
```

Or, this more compact one:

```
        return if (!networkUtilsWrapper.isNetworkAvailable()) {
            emit(NetworkUnavailable)
            true
        } else false
```

ParaskP7 replied:
This is done here: 5cc156dd031b34c4f95c64055633cb04e672c36e

I went with the more compact approach, let me know if you have another idea. 🙏

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/jetpack/restore/usecases/GetRestoreStatusUseCase.kt:
```diff
@@ -1,65 +1,100 @@
+        val result = activityLogStore.fetchActivitiesRewind(FetchRewindStatePayload(site))
+        if (result.isError) {
+            emit(RemoteRequestFailure)
+            return true
```
❓ It feels a bit unintuitive to me that the method returns "true" on failure and "false" on success. Could you please elaborate why you made this decision?

ParaskP7 replied:
I agree @malinajirka , it felt unintuitive to me too.

I was just extracting this logic to remove duplication and this was the default structure the IDE came up automatically. I then was thinking a bit about it and didn't like it but couldn't come up with a better naming, then I moved forward, thanks for bringing that up again.

What if I rename that function to `fetchActivitiesRewindAndReturnOnError(...)` or revert the logic and name it as `isFetchActivitiesRewindSuccessful(...)` to match the `isNetworkAvailable()`?

ParaskP7 replied:
This is done here: 05e4b5bfcb78ead282be53a1dcf6f93d2abbfb3f

I went with the `isFetchActivitiesRewindSuccessful(...)` approach, let me know if you have another idea. 🙏

malinajirka replied:
Thanks for the changes! I'm still on the edge - when I see `isFetchActivitiesRewindSuccessful` I don't expect the method will actually hit the server. What if we just named it "fetchActivitiesRewind" and return "true" on success and "false" on failure? Wdyt?

ParaskP7 replied:
I agree. 🙏

ParaskP7 replied:
This is done here: 4de839318642e1b3beda35bda2c0c45bf24f2072

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/jetpack/restore/usecases/PostRestoreUseCase.kt:
```diff
@@ -28,7 +28,7 @@ class PostRestoreUseCase @Inject constructor(
         rewindId: String,
         site: SiteModel,
-        types: RewindRequestTypes
+        types: RewindRequestTypes? = null
```
❓ Could we perhaps pass RewindRequestTypes(true,true,...) instead of introducing a nullable type here? I realize it'll slightly change the current behavior (the request will contain the types instead of an empty request) but I think it's worth it. Wdyt?

ParaskP7 replied:
Yeap, another good catch @malinajirka , you, you, you are good! 😄 

So, the reason why I decided to go with `null` instead of `RewindRequestType(true, true, ...)` is because the `RewindStatusService.rewind(...)` function is also doing the same, so I didn't want to change the behavior as I wasn't sure if that will end up being a functional change. See `val rewindResult = activityLogStore.rewind(RewindPayload(site, rewindId))`.

On the `FluxC` level passing `null` versus `types` is different, see `ActivityLogRestClient.rewind(...)` function and more specifically this block of code:
```
        val typesBody = if (types != null) {
            mapOf("types" to types)
        } else {
            mapOf()
        }
```

So on what case you would add an `empty map`, while on the other case you would add a `map of types`.

Wdyt?

zwarm replied:
iirc, the empty map is acceptable in the API and is the default. I'll have to go back and test to be positive, that was so long ago.

malinajirka replied:
Yeah, I realize it'd slightly change the behavior but from my understanding the result will be the same. Having said that, let's not worry about that now and lets just stick to the nullable type. Thanks!

malinajirka reviewed (changes requested):
Really great job @ParaskP7 🥇 ! The logic is incomparably easier to follow now and the concerns are nicely separated ❤️ . I have left a few comments, but nothing vital. I tested the behavior with feature flags on/off and it worked pretty well - see possible issues below.

Note: This PR is quite huge and I found it difficult to review. You did a great job documenting your every step, but review requires focus and being focused while reviewing 70+ commits quite tricky (even if each commit would be extremely simply, which some of them aren't). Tbh I gave up the review-by-commit and I reviewed just the changes - which turned out to be ok, since most of the changes are in tests which don't require such a focused review.
Please always consider splitting PRs into smaller chunks - each of the PRs can be merged into a working branch and you can request that only the final PR needs to be tested. Thanks!


Possible issues
1. ⚠️ When I turn off all feature flags -> open Activity Log -> click on Restore -> it takes +- 5s before the item at the top with progress appears. Before this PR, the item would appear right away.
2. When I enable "restore" feature flag -> open Activity Log -> Click on "Restore" -> Click on "Restore site" -> Click on "Confirm restore site" -> click on "Ok notify me" -> the progress item doesn't appear, unless I pull down to refresh. (I think this behavior was the same even before this PR)

ParaskP7 commented:
> Really great job @ParaskP7 🥇 ! The logic is incomparably easier to follow now and the concerns are nicely separated ❤️ . I have left a few comments, but nothing vital. I tested the behavior with feature flags on/off and it worked pretty well - see possible issues below.

Thanks for the review @malinajirka ! 🌟 

> Note: This PR is quite huge and I found it difficult to review. You did a great job documenting your every step, but review requires focus and being focused while reviewing 70+ commits quite tricky (even if each commit would be extremely simply, which some of them aren't). Tbh I gave up the review-by-commit and I reviewed just the changes - which turned out to be ok, since most of the changes are in tests which don't require such a focused review.

Yes, I could probably merge some of the commits together, like the import commits. Even with test refactoring, maybe there I could have been more relaxed as well, will consider it for next time. Many thanks! 🙏 

> Please always consider splitting PRs into smaller chunks - each of the PRs can be merged into a working branch and you can request that only the final PR needs to be tested. Thanks!

I really considered that, it is just that the overall solution was keep getting a bit more involved as I was progressing with it. Now that you know what was involved, how would you split the PR further, can you make a suggestion? 🙏 

> ⚠️ When I turn off all feature flags -> open Activity Log -> click on Restore -> it takes +- 5s before the item at the top with progress appears. Before this PR, the item would appear right away.

Yes, nice catch, I should have mentioned that in the description as well. This is expectable because of the way the use case works and the 5' seconds delay. I didn't want to complicate things further thus I decided to not deal with it now. Do you this it is fine if you launch with that as an MVP and improve later. I'll then add a task on the parent ticket. 🌟 

> When I enable "restore" feature flag -> open Activity Log -> Click on "Restore" -> Click on "Restore site" -> Click on "Confirm restore site" -> click on "Ok notify me" -> the progress item doesn't appear, unless I pull down to refresh. (I think this behavior was the same even before this PR)

Yes, this is going to be part of my next PR. The current `develop` version is also not working for the `Ok, notify me!` or when going back manually. See `Hook wizard back into Activity Log List` task on the parent ticket. 🌟

ParaskP7 committed:
Refactor: Rename is network unavailable to available on restore use case

ParaskP7 committed:
Refactor: Rename generic date to published for restore states and event

ParaskP7 committed:
Refactor: Combine two returns into one on restore use case

ParaskP7 committed:
Refactor: Rename fetch activities rewind on restore use case

zwarm commented:
@ParaskP7 - So much easier to understand, very nicely done! 🎆 

I've have a few takeaways that I will be refactoring in GetBackupDownloadStatus based off of the comments here:
- Move the flowOn operator/dispatcher to use case
- Replace Date with published in the states
- IsNetworkUnavailable to NetworkAvailable
However, I don't know if you have the backup integration waiting in a separate PR. Let me know and I won't touch them.

One comment not related to the code in this PR, but related to the use case.
-I still have an item on my list to add 3X error retry on fetch error which may alter `if (!isFetchActivitiesRewindSuccessful(site)) return@flow` - not truly a success, but not a failure ... don't want to exit flow, but restart the loop and not go to the next step. May be nice to sync up on this one. (I have already added the retry to the backup use case)

malinajirka commented:
> I really considered that, it is just that the overall solution was keep getting a bit more involved as I was progressing with it. Now that you know what was involved, how would you split the PR further, can you make a suggestion? 🙏

I usually prefer to have PRs with a single focus and this PR was making a lot of improvements not directly related to the task. I really love these improvements. I was wondering if we could have for example created a separate PR for test refactoring + the new tests added before you actually started working on the changes. Perhaps it wasn't feasible, it's hard to tell as a reviewer. Thank you so much for considering it! 

> Yes, nice catch, I should have mentioned that in the description as well. This is expectable because of the way the use case works and the 5' seconds delay. I didn't want to complicate things further thus I decided to not deal with it now. Do you this it is fine if you launch with that as an MVP and improve later. I'll then add a task on the parent ticket. 🌟

👍 

> Yes, this is going to be part of my next PR. The current develop version is also not working for the Ok, notify me! or when going back manually. See Hook wizard back into Activity Log List task on the parent ticket. 🌟

❤️

ashiagr commented on WordPress/src/test/java/org/wordpress/android/viewmodel/activitylog/ActivityLogViewModelTest.kt:
```diff
@@ -662,30 +568,568 @@ class ActivityLogViewModelTest {
-        )
-        list.add(activity)
-        list.add(activity.copy(rewindable = false))
+    private fun event() = ActivityLogListItem.Event(
```
❓: Functions with no arguments might be interchangeable with read-only properties. Any reason why we used a function `event()` here? Might help me understand when to prefer one over another.

ParaskP7 replied:
Thanks for the question @ashiagr ! And what a great one indeed. 🌟

You will understand the answer quickly if you check the other such function further above, the `activity()` one, which accept parameters, which all of them default to some value. So, in the end you can use this function as `activity()` and get the default behavior of the (data) class, or update say the `rewindable` to `false` and get an activity that is not rewindable. With the `event()` I went with the same approach, but didn't need to add any variables (yet).

To be honest, there is a longer story to that. This is actually a `stub/fake` pattern that can be applied to any codebase, in terms of testing, where ALL those suchlike function can be created in one place, with all its parameters being available from the function, with default, so that whenever a developer needs to create say an `activity` for their test, they would just import this helper function from that common place and right away use it as `activity()`, without the need to do all this boilerplate setup over and over again. This really removes the clutter in setting up tests by a lot. Also, this helps removing the need to use `@Mock` on a `Site` class, which is not recommended, and have a common `site()` test helper function to have a site created as a stub/fake and not as a mock.

Does the above make sense or am I mumbling again? 😅

ashiagr replied:
Thank you for an elaborate answer 🙏, it does make sense!

ashiagr reviewed (approved):
@ParaskP7 , I reviewed each commit and appreciate the level of effort put into building the solution. I really liked the refactoring effort like
- re-ordering functions close to the places they're used
- creating private group
- extracting smaller functions out of bigger ones, increasing readability

Great job!

I agree with @malinajirka that this PR could've been split into smaller PRs. But I also understand that sometimes it is difficult to do that while being so much involved in building a solution. It happens to everyone of us. Nice thing is commits are split into separate ones specific to their purpose. We might consider cherry picking some of the similar commits (like for tests refactoring) and creating a separate PR out of it. It might be too much of an effort, and as always we have to find a balance. 

I just left one question. Approving it from side.

ParaskP7 committed:
Refactor: Rename fetch activities rewind on restore use case

ParaskP7 commented:
👋 @zwarm !

> @ParaskP7 - So much easier to understand, very nicely done! 🎆

Thank YOU! ❤️ 

> However, I don't know if you have the backup integration waiting in a separate PR. Let me know and I won't touch them.

I am going to create 2 PRs today, one for the restore integration and one for the backup integration. It is probably better if I do the change on the `GetBackupDownloadStatusUseCase` myself so that you can I don't block you and you can focus on other things, wdyt? 🙏 

> I still have an item on my list to add 3X error retry on fetch error which may alter if (!isFetchActivitiesRewindSuccessful(site)) return@flow - not truly a success, but not a failure ... don't want to exit flow, but restart the loop and not go to the next step. May be nice to sync up on this one. (I have already added the retry to the backup use case)

I remember the 3x error retry login on the `GetBackupDownloadStatusUseCase`. Feel free to introduce a similar logic for the `GetRestoreStatusUseCase` when this PR is merged. And definitely, let's sync! 🌟

ParaskP7 commented:
👋 @malinajirka !

> I usually prefer to have PRs with a single focus and this PR was making a lot of improvements not directly related to the task. I really love these improvements. I was wondering if we could have for example created a separate PR for test refactoring + the new tests added before you actually started working on the changes. Perhaps it wasn't feasible, it's hard to tell as a reviewer. Thank you so much for considering it!

I so much agree, and apologies for making the move and created one PR instead of two, I could definitely do what you suggested (refactor + tests) and then the actual solution. Please expect that next time. I probably got caught in the fever of delivering since I have been delaying this solution more than I expected (or wanted). 🙏

ParaskP7 commented:
👋 @ashiagr !

> @ParaskP7 , I reviewed each commit and appreciate the level of effort put into building the solution. I really liked the refactoring effort like

❤️ 

> ? I agree with @malinajirka that this PR could've been split into smaller PRs. But I also understand that sometimes it is difficult to do that while being so much involved in building a solution. It happens to everyone of us. Nice thing is commits are split into separate ones specific to their purpose. We might consider cherry picking some of the similar commits (like for tests refactoring) and creating a separate PR out of it. It might be too much of an effort, and as always we have to find a balance.

Yes, I totally agree with you all, I should have split this PR at least into two (refactor + tests). Thanks so much such suggestions. Please expect me to do that next time. 🙏

malinajirka commented:
> I so much agree, and apologies for making the move and created one PR instead of two, I could definitely do what you suggested (refactor + tests) and then the actual solution. Please expect that next time. I probably got caught in the fever of delivering since I have been delaying this solution more than I expected (or wanted). 🙏

Please don't apologize :). It's very difficult to find the balance before you start working on a task. As you progress it becomes more and more difficult to split the work up. Thank you for all the effort you put into documenting your every step and for the great detailed PR description! It helped all of us ;)

ParaskP7 commented:
👋 @malinajirka !

> Yes, this is going to be part of my next PR. The current develop version is also not working for the Ok, notify me! or when going back manually. See Hook wizard back into Activity Log List task on the parent ticket. 🌟

The PR I was referring to is now ready, you can check it really quick if you are curious: #13914
